File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Factory-method/Factory-method.js
class UserInfo {
  constructor(username, password, { key, address } = {}) {
    this.username = username;
    this.password = password;
    this.key = key;
    this.address = address;
  }
}

class UserFactory {
  static create(role, info) {
    switch (role) {
      case "admin":
        return new Admin(info);
      case "seller":
        return new Seller(info);
      default:
        throw new Error("Invalid role");
    }
  }
}

class Admin extends UserInfo {
  constructor(info) {
    super(info.username, info.password, info);
  }
}

class Seller extends UserInfo {
  constructor(info) {
    super(info.username, info.password, info);
  }
}

const adminJs = UserFactory.create("admin", {
  username: "mahdi",
  password: "1234",
  key: "secret_admin",
});

console.log("admin:", adminJs);

File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Factory-method/Factory-method.py
class UserInfo:
    def __init__ (self,username,password,key,address):
        self.username = username
        self.password = password
        self.key = key
        self.address = address
class UserFactory:
     def  create_user(self,role,info):
        creators = {
            "admin": Admin,
            "seller": Seller,
        }
        creator = creators.get(role)
        return creator(info)
class Admin(UserInfo):
    def __init__(self,info):
        super().__init__(**info)
class Seller(UserInfo):
    def __init__(self,info):
        super().__init__(**info)

admin = UserFactory().create_user("admin", {
    "username": "mahdi",
    "password": "2312",
    "key": "my_secret",
})
print("admin:", admin)